src/utils/multipart-parser.js

```diff
@@ -0,0 +1,440 @@
+					state = S.PART_DATA_START;
+					break;
+				case S.PART_DATA_START:
+					state = S.PART_DATA;
+					mark('partData');
+					// fallsthrough
+				case S.PART_DATA:
+					previousIndex = index;
+
+					if (index === 0) {
+						// boyer-moore derrived algorithm to safely skip non-boundary data
+						i += boundaryEnd;
+						while (i < bufferLength && !(ui8a[i] in boundaryChars)) {
+							i += boundaryLength;
+						}
+
+						i -= boundaryEnd;
+						c = ui8a[i];
+					}
+
+					if (index < boundary.length) {
+						if (boundary[index] === c) {
+							if (index === 0) {
+								dataCallback('partData', true);
+							}
+
+							index++;
+						} else {
+							index = 0;
+						}
+					} else if (index === boundary.length) {
+						index++;
+						if (c === CR) {
+							// CR = part boundary
+							flags |= F.PART_BOUNDARY;
+						} else if (c === HYPHEN) {
+							// HYPHEN = end boundary
+							flags |= F.LAST_BOUNDARY;
+						} else {
+							index = 0;
+						}
+					} else if (index - 1 === boundary.length) {
+						if (flags & F.PART_BOUNDARY) {
+							index = 0;
+							if (c === LF) {
+								// unset the PART_BOUNDARY flag
+								flags &= ~F.PART_BOUNDARY;
+								callback('partEnd');
+								callback('partBegin');
+								state = S.HEADER_FIELD_START;
+								break;
+							}
+						} else if (flags & F.LAST_BOUNDARY) {
+							if (c === HYPHEN) {
+								callback('partEnd');
+								callback('end');
+								state = S.END;
+								flags = 0;
+							} else {
+								index = 0;
+							}
+						} else {
+							index = 0;
+						}
+					}
+
+					if (index > 0) {
+						// when matching a possible boundary, keep a lookbehind reference
+						// in case it turns out to be a false lead
+						lookbehind[index - 1] = c;
+					} else if (previousIndex > 0) {
+						// if our boundary turned out to be rubbish, the captured lookbehind
+						// belongs to partData
+						const _lookbehind = new Uint8Array(lookbehind.buffer, lookbehind.byteOffset, lookbehind.byteLength);
+						callback('partData', 0, previousIndex, _lookbehind);
+						previousIndex = 0;
+						mark('partData');
+
+						// reconsider the current character even so it interrupted the sequence
+						// it could be the beginning of a new sequence
+						i--;
+					}
+
+					break;
+				case S.END:
+					break;
+				default:
+					return i;
+			}
+		}
+
+		dataCallback('headerField');
+		dataCallback('headerValue');
+		dataCallback('partData');
+
+		this.index = index;
+		this.state = state;
+		this.flags = flags;
+
+		return length_;
+	}
+
+	end() {
+		function callback(self, name) {
+			const callbackSymbol = 'on' + name.slice(0, 1).toUpperCase() + name.slice(1);
+			if (callbackSymbol in self) {
+				self[callbackSymbol]();
+			}
+		}
+
+		if ((this.state === S.HEADER_FIELD_START && this.index === 0) ||
+			(this.state === S.PART_DATA && this.index === this.boundary.length)) {
+			callback(this, 'partEnd');
+			callback(this, 'end');
```

Comment:
`callbackSymbol in self` will try to call `undefined` if the user sets e.g. `onPartEnd` to `undefined`, what do you think about checking `self[callbackSymbol] != null` instead (which would then allow `= undefined`, `= null`, or `delete` to clear it)?

-----

Generally, I think that doing dynamic lookups to make the call site cleaner increases the cognitive load of reading the code. Personally, the most straightforward thing that I would have liked would have been `this.onPartEnd?.()`, but since we are still supporting Node.js 12 we can't do that 😅. Still, I think that something like this would still make the code easier to read:

```suggestion
		function callback(fn) {
			if (fn != null) fn();
		}

		if ((this.state === S.HEADER_FIELD_START && this.index === 0) ||
			(this.state === S.PART_DATA && this.index === this.boundary.length)) {
			callback(this.onPartEnd);
			callback(this.onEnd);
```

and then maybe hoist up the function and call it something generic:

```js
function callIfNonNullish(fn, ...args) {
  if (fn != null) fn(...args)
}
```

I think that this avoids one level of indirection which might make it a bit harder to read. Specifically in this `end` function it isn't that much of a problem of course, but I had some hard times following the `callback` and `callbackData` in the other function (will put a comment there after I have read thru and understood all of that code, just started here ☺️)

Reply:
the hole multipart decoder is just a internal core part and never exposed to the end user, but i will take your suggestions

Reply:
> this.onPartEnd?.(), but since we are still supporting Node.js 12 we can't do that

yea, would also like to do that... for now i have done something like `x && x()` in other packages...